elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +306,319 @@ private Object fetchRelationship(Environment context) {
         }
     }
 
-    /** stub code **/
-    private Object replaceObjects(Environment context) { return null; }
+    /**
+     * handle UPSERT operation
+     * @param context environment encapsulating graphQL's request environment
+     * @return list of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
+
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
+
+        Set upsertOps = new HashSet();
+        /* upsert */
+        for(Map<String, Object> input : context.data.get()) {
+            upsertOps.addAll(graphWalker(input, this::upsertObject, Optional.ofNullable(context.parentResource),
+                    context.requestScope, context.field.getName()));
+        }
+
+        /* fixup relationships */
+        for(Map<String, Object> input : context.data.get()) {
+            graphWalker(input, this::updateRelationship, Optional.ofNullable(context.parentResource),
+                    context.requestScope, context.field.getName());
+        }
+
+        return upsertOps;
+    }
+
+    /**
+     * utility class to support storing a triplet on each node of the queue used while upserting
+     */
+    private class Triplet {
+        private Map<String, Object> input;
+        private Optional<PersistentResource> parentResource;
+        private String fieldName;
+
+        /**
+         * class constructor
+         * @param input data input
+         * @param parentResource parent
+         * @param fieldName field name
+         */
+        private Triplet(Map<String, Object> input, Optional<PersistentResource> parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * A function to handle upserting (update/create) objects.
+     */
+    @FunctionalInterface
+    private interface Executor<T> {
+        T execute(Map<String, Object> input, String id,
+                                   RequestScope requestScope, Class<?> entityClass,
+                                   String fieldName, Optional<PersistentResource> parent);
+    }
+
+    /**
+     * Forms the graph from data {@param input} and upsert's {@param function} all the nodes
+     * @param input input data
+     * @param function function to process nodes
+     * @param parent parent resource
+     * @param requestScope request scope
+     * @param fieldName graphql field name
+     * @return set of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> graphWalker(Map<String, Object> input, Executor function,
```

Comment:
This will be hard to explain in comments, but there is a much simpler way to do this.  This walker is attempting to  walk each relationship build tuples like (E1, E2, R) where E1 and E2 are parent and child entities and R represents a relationship between them.  Trying to fixup the object tree this way is problematic because it is doing it at too fine a grain.

What you really need is no tuple at all.   The walker should queue up just the entity (E1).

The driver function will call the walker twice.  The first time, the execute function will just
create the entity and set its attributes.  IT WILL NOT try to set any relationships (including fixing up its parent).  

The second walker also visits each entity.  For each entity, it will navigate the provided relationships.  For each relationship, it will attempt to load all the objects provided in the data.  It will then attempt to REPLACE the entire relationship.

There is one final steps, the driver function still needs to link the top level objects to its parent.  That doesn't happen in the walker execute functions though.  It happens in the driver.